### PR TITLE
Custom Shortcut update

### PR DESCRIPTION
**Pull request by VictoriousRaptor**
Implemented changes required in #930.

### CODE REVIEW

**Comment by onesounds** [comment]
Hi, Welcome on board.
1. I added one item and tested.
2. try to typing {clipboard} <- error happened. 
3. Close -> Rebuild -> error in jsonstorage.cs. and stucked here.

**Comment by VictoriousRaptor** [comment]
Current problem

> Hi, Welcome on board.
> 
>     1. I added one item and tested.
> 
>     2. try to typing {clipboard} <- error happened.
> 
>     3. Close -> Rebuild -> error in jsonstorage.cs. and stucked here.
> 
> 
> 
> 
> Setting. 

Sorry about that. I've noticed this bug right after creating this pr. Try pull it again. It should be fixed now.

**Comment by VictoriousRaptor** [comment]
current problem:
- [ ] The design is not elegant when adding/deleting/editing shortcut
- [ ] May trigger an exception when typing {clipboard} with a plugin using action keyword "{" (VS Code Workspace in my case)

**Comment by onesounds** [comment]
Don't be sorry. I hope you're not stressed. I saw a lot of people who were stressed out and exhausted by the fertilization work.

Anyway, before issue is fixed. and then...
- Edit not applied.
- Done made error in last item.

Animation6

**Comment by taooceros** [comment]
Well my original design doesn't really want user to type {clipboard} (isn't that too messy to type? But leave that as a place holder for the expansion.

**Comment by VictoriousRaptor** [comment]
> I skim through the code and i couldn't understand how shall plugin provide dynamic shortcuts? Could you explain a little about how it really works?

I don't have any good idea so like you said, I leave it to the future updates.

**Comment by VictoriousRaptor** [comment]
> Don't be sorry. I hope you're not stressed. I saw a lot of people who were stressed out and exhausted by the fertilization work.
> 
> Anyway, before issue is fixed. and then...
> 
>     * Edit not applied.
> 
>     * Done made error in last item.
> 
> 
> Animation6 [ Animation6 ](https://user-images.githubusercontent.com/6903107/195602301-da6acc14-1f42-430c-8831-a60e00ffe30b.gif) [ ](https://user-images.githubusercontent.com/6903107/195602301-da6acc14-1f42-430c-8831-a60e00ffe30b.gif)

I'm working on it.

**Comment by VictoriousRaptor** [comment]
> > Don't be sorry. I hope you're not stressed. I saw a lot of people who were stressed out and exhausted by the fertilization work.
> > Anyway, before issue is fixed. and then...
> > ```
> > * Edit not applied.
> > 
> > * Done made error in last item.
> > ```
> 
> I'm working on it.

Should be fixed now.

File: Flow.Launcher/ViewModel/SettingWindowViewModel.cs [hunk 1]
```diff
@@ -44,6 +44,7 @@ public SettingWindowViewModel(Updater updater, IPortable portable)
                         break;
                 }
             };
+            ShortCuts = new ObservableCollection<CustomShortcutModel>(Settings.CustomShortcuts.Union(Settings.BuiltinShortcuts));
```
Comment (taooceros) [inline Flow.Launcher/ViewModel/SettingWindowViewModel.cs:47]: I kindly don't like displaying them together in one control....Let's move them to separate control.

Comment (taooceros) [reply Flow.Launcher/ViewModel/SettingWindowViewModel.cs:47]: Also build in shortcut has a different replace policy (no prefix `@` is needed), so shall them share the same name (shortcut)?

Though, I really appreciate your work to help improve the Extensibility. Thanks

File: Flow.Launcher/SettingWindow.xaml.cs [hunk 2]
```diff
@@ -386,33 +394,41 @@ private void OnDeleteCustomShortCutClick(object sender, RoutedEventArgs e)
...
-                var shortcutSettingWindow = new CustomShortcutSetting(item.Value);
-                if (shortcutSettingWindow.ShowDialog() == true)
-                {
-                    settings.ShortCuts[viewModel.SelectCustomShortcutIndex.Value] = shortcutSettingWindow.ShortCut;
-                }
+                MessageBox.Show(InternationalizationManager.Instance.GetTranslation("pleaseSelectAnItem"));
+                return;
             }
-            else
+            else if (!item.CanBeEdited)
```
Comment (taooceros) [inline Flow.Launcher/SettingWindow.xaml.cs:409]: Let's remove this check. The buildin shortcut is more like a variable that is dynamic. We may want to display them with their key and description (not just the value).

Although I am thinking there may be snippet plugins somewhat like the vscode that only provides a bunch of snippets, but I also think the description can be the value. Therefore, let's make them separate and be clear about everything.

**Review by taooceros — COMMENTED** [review]
It's a very nice improvement! Thank you so much. I am saying we may want to use another control to display the buildin shortcut. However, if you don't want to touch the UI code, we can just fix other concern and leave all these UI works to @onesounds 😝.

File: Flow.Launcher/ViewModel/SettingWindowViewModel.cs (hunk 1 above)
Comment (VictoriousRaptor) [reply Flow.Launcher/ViewModel/SettingWindowViewModel.cs:47]: I separated builtin and custom shortcuts. What's your idea about displaying bulit in shortcuts? Putting them in a list similar to the existing one?

Comment (VictoriousRaptor) [reply Flow.Launcher/ViewModel/SettingWindowViewModel.cs]: Image

Like this?

Comment (taooceros) [reply Flow.Launcher/ViewModel/SettingWindowViewModel.cs]: Yeah that shall be good

Comment (VictoriousRaptor) [reply Flow.Launcher/ViewModel/SettingWindowViewModel.cs]: any other changes needed?

**Comment by taooceros** [comment]
Thanks that's pretty great! We can leave some design work to @onesounds 😝

**Pull request merged**